ct: Add `write_pipeline` and `throttler` components

Lazin opened the pull request:
Currently, the batcher is unnecessary complicated component. It's responsible for storing `write_request` instances, performing housekeeping (expiring), and uploading.

I wanted to add throttling to the batcher to avoid overloading Redpanda and keep memory used by in-flight `write_request` instances in check. With current architecture all this functionality should be added to the batcher itself because it maintains the list of write requests. Other aspects like caching and load balancing could also be added to the batcher complicating it even further.

To avoid this I refactored the `batcher` and split it into two components, the `batcher` and the `write_pipeline`. The `write_pipeline` is accepting and maintaining `write_request` instances. It can expire old write requests and can return size limited set of in-flight write requests in both FIFO and LIFO orders. It has an interface that allows the caller to subscribe to certain events (new stuff is added to the pipeline). The `event_filter` component implements the subscription mechanism.

The `batcher` component now uses the `write_pipeline` to upload the data. It subscribes to events and detects when there is enough data to start L0 upload (or enough time have passed). So the functionality of the `batcher` is not limited to two things: interacting with the `write_pipeline` and uploading the data.

Previously, we could just produce data to the batcher directly:
```
auto placeholders = co_await batcher.write_and_debounce(std::move(reader), timeout);
```

With this reorg we will have to connect batcher to the pipeline and then produce data to the pipeline:
```
core::write_pipeline pipeline;
batcher batcher(pipeline, bucket, mock);
...
auto placeholders = co_await pipeline.write_and_debounce(std::move(reader), timeout);
```

This PR also adds `throttler` component. The `throttler` is similar to batcher. It subscribes to `write_pipeline` and inspects new write requests. If the throughput limit is exceeded it moves newest write requests out of the pipeline for some period of time and returns them back (or discards them if they are discarded). If the new write request exceeds the memory limit it discards the write request and acknowledges the client using the `slow_down` error code.

Similar modular approach could also be used for the read path. The main goal of this is to make the write path as modular as possible so it could be developed  and tested in parallel. Things like in-memory caching and load balancing should also be added as separate services on top of the `write_pipeline`.

This is a work in progress. Some unit-tests in the `throttler` are missing.
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

vbotbuildovich commented:
<!--UNIT FAILURES-->
<!--94634e4d-8284-4ecc-90b7-ad0b10831b91-->
<!--body-->
the below tests from https://buildkite.com/redpanda/redpanda/builds/57534#0192f76d-ffbe-4623-9477-06fa42b73564 have failed and will be retried
```
catalog_schema_manager_rpunit
```
<!--/body-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--54ed1f3a-fd70-4fda-a661-70fe01cdcc13-->
<!--RETRIED FAILURES-->
<!--1922fd88-4844-4a51-8e02-24fbc65d1354-->
<!--RETRIED FAILURES-->
<!--a37987d2-eafc-4188-a566-d6f5c481d4a3-->
<!--RETRIED FAILURES-->
<!--50fa616a-5f9b-4773-aaa1-d068640de236-->
<!--RETRIED FAILURES-->
<!--aa1503ea-3045-43b9-8783-a60e15689a88-->
<!--RETRIED FAILURES-->
<!--26fb3b72-dce0-4fbe-9739-7ff3693252ab-->
<!--RETRIED FAILURES-->
<!--6617c84a-df76-411e-bc41-0b5d841a55c3-->
<!--RETRIED FAILURES-->
<!--0b03ebf3-c503-4221-afa8-f8306678e48c-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/57699#0193033e-c620-45df-862d-811a8089e843
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/57929#01931d18-5c5a-45ed-ac35-3f30e026fd6a
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/57929#01931d18-5c5d-4bc0-b8c1-9b29fa406cda
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/57958#01932124-7027-47b1-a560-fcf94fbc042a
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/58020#019326af-93a8-4ae7-a406-ac868c92c8a0
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/58124#01933118-e27e-45fc-9024-401225a6691c
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/58124#01933134-23e7-4540-a9c4-88e365a0b770
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59056#01937afc-500a-45ed-b3ce-4ce4c32e84aa
<!--/body-->

nvartolomei commented on src/v/cloud_topics/core/BUILD:
```diff
@@ -59,6 +59,20 @@ redpanda_cc_library(
     ],
```
nit: this commit should be squashed with the one adding event_filter.h

nvartolomei commented on src/v/cloud_topics/core/pipeline_stage.h:
```diff
@@ -0,0 +1,23 @@
+namespace experimental::cloud_topics::core {
+
+/// Processing stage id
+using pipeline_stage = named_type<int, struct _pipeline_stage_tag>;
```
IMHO this should be a class with a private constructor with pipeline declared as friend.

Much clearer indicates that it is a opaque type and no one should mess with it. named_type allows reading it, modifying it, constructing arbitrary values.

Also, should be squshed with the commit which uses this type.

nvartolomei replied:
> Much clearer indicates that it is a opaque type and no one should mess with it. named_type allows reading it, modifying it, constructing arbitrary values.

Lazin replied:
the question was about squashing

Lazin replied:
I understand the motivation behind opaque type. It's just not obvious to me that we should do this. THe issue is quite minor. The code is used only internally. There will be like 3 or 4 pipeline stages inside the cloud_topics package and no external users of this stuff.

Lazin replied:
I also extremely dislike friendship over long distances. If something lives in a different module it shouldn't be a friend. I'll think about some other approach though.

nvartolomei commented on src/v/cloud_topics/core/BUILD:
```diff
@@ -94,6 +94,7 @@ redpanda_cc_library(
         "//src/v/bytes",
```
Please squash with the commit which defines the write_pipeline. I guess that commit failed bazel build at all now

nvartolomei commented on src/v/cloud_topics/throttler/tests/throttler_test.cc:
```diff
@@ -0,0 +1,546 @@
+// Copyright 2024 Redpanda Data, Inc.
```
Tests together with code 🙏

nvartolomei replied:
https://github.com/redpanda-data/redpanda/pull/23919#discussion_r1834137524

nvartolomei commented on src/v/cloud_topics/core/tests/write_pipeline_test.cc:
```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 Redpanda Data, Inc.
```
Also when reviewing commit by commit I have now to jump between 2 commits to understand the intention behind particular code.

nvartolomei commented on src/v/cloud_topics/core/event_filter.h:
```diff
@@ -0,0 +1,127 @@
+/// by events.
+/// E.g. the background upload loop in the batcher can start L0
+/// upload early if enough data is accumulated.
+template<class Clock = ss::lowres_clock>
```
I wonder if it is possible to templatize like for example `event_filter<event_type::write_request>`? Don't have enough metaprogramming experience to asses whether this is possible without digging deeper.

this way when you co_await c.subscribe(...) you won't need a switch or when you subscribe with timeout you would need to switch but only on write_request and timeout rather than on write requests, shutdown (this can also be just a timeout event) so in the normal case you could have 2 events.

Or just propagate timeout, shutdown as exception and have no switch at all.

nvartolomei commented on src/v/cloud_topics/throttler/throttler.cc:
```diff
@@ -0,0 +1,221 @@
+namespace experimental::cloud_topics {
+
+template<class Clock>
+throttler<Clock>::throttler(
```
This is interesting but have 2 big questions:

1. Why throttler needs to copy requests internally and manage them instead of not progressing them to the next pipeline stage?
2. Can throttler introduce back-pressure? I believe that would be desirable but in this design it doesn't seem possible. Pipeline will keep accepting requests in _pending is if throttler is able to keep up (not guaranteed)? then it will drop requests when memory is full but if it doesn't keep up what happens?
3. In the current design with background threads there seems to be quite a bit of overhead. Can't we make items move through the pipeline eagerly? Rather than have "pipeline stages" "pull" elements make "pipeline" push elements to the first stage and the first stage will either drop/hold/push further the element. Out of the box we get backpressure support as well?. Reminds me of this talk https://dsdsd.da.cwi.nl/slides/dsdsd-duckdb-push-based-execution.pdf.

nvartolomei commented on src/v/cloud_topics/core/write_request.cc:
```diff
@@ -8,24 +8,25 @@
  */
 
-#include "cloud_topics/batcher/write_request.h"
+#include "cloud_topics/core/write_request.h"
```
there is no reason not make linear (commit by commit) reviewing easy for the reader and later when blaming

i don't want to walk commits and play with diffing various combinations when it costs nothing to make the entire change in a single commit

nvartolomei commented on src/v/cloud_topics/core/README.md:
```diff
@@ -0,0 +1,8 @@
+The `core` contains utilities shared by read path, write path and resource balancer.
```
> also with bazel it feels much nicer to have a lot of fine grained libs inside the sub-project dir. It encourages to do this

then why not do this insetad of slapping {write,read}_request, {write,read}_pipeline in a generically named package called "core"? :)

> It's not supposed to be a part of the public api

by public api i mean whether these classes will be constructed from outside of cloud_topics package (i assumed batcher will be and hence pipelines too)

nvartolomei commented on src/v/cloud_topics/core/tests/write_pipeline_test.cc:
```diff
@@ -0,0 +1,122 @@
+
+    auto res = pipeline.get_write_requests(
+      1, cloud_topics::core::initial_pipeline_stage);
+    ASSERT_TRUE_CORO(res.complete);
```
I find it easier to ensure basic things are correct rather than be surprised later with bugs. I understand the argument and I find it acceptable if there is good degree of confidence that the alternative code paths are covered by some other tests.

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--f7a7b18f-e911-451a-b008-81bd40ceee59-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/58020#019326af-93ad-49d8-816e-ae80d6f1e04c:
```
"rptest.tests.archive_retention_test.CloudArchiveRetentionTest.test_delete.cloud_storage_type=CloudStorageType.ABS.retention_type=retention.ms"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/58020-->
<!--d552f387-a62e-4284-8b32-40d3a736eb2f-->
<!--title-->
## Retry command for [Build#58020](https://buildkite.com/redpanda/redpanda/builds/58020)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.ms"}
```
<!--/body-->

Lazin commented:
/ci-repeat 1
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.ms"}

Lazin commented:
The `throttler` component was divided into two. The `throttler` now handles only throughput. The `evictor` removes write requests if memory limit was reached.

Lazin commented:
Changed the code a bit. Removed the `evictor` and replaced it by the semaphore inside the pipeline. Every write request has to acquire the units from this semaphore first.

nvartolomei commented on src/v/cloud_topics/core/event_filter.h:
```diff
@@ -0,0 +1,127 @@
+        _expiry->arm(deadline);
+    }
+
+    void cancel() {
```
maybe also

```cpp
if (_expiry.has_value()) {
  _expiry.cancel();
}
```

Lazin replied:
d-tor of the timer disarms it

nvartolomei commented on src/v/cloud_topics/core/tests/event_filter_test.cc:
```diff
@@ -0,0 +1,183 @@
+      stage,
+      ss::lowres_clock::now() + 1ms);
+    auto sub = pipeline.subscribe(flt);
+    co_await ss::sleep(10ms);
```
Why you are not using manual clocks here?

Lazin replied:
why not?

nvartolomei commented on src/v/cloud_topics/core/write_pipeline.h:
```diff
@@ -0,0 +1,148 @@
+                checked<write_request_process_result, errc> r = fn(req);
+                if (r.has_error()) {
+                    // Drop write request using the error code from the result
+                    req.stage = _stages.next_stage(req.stage);
```
Do we need to advance the stage in case of error? I'd expect not to 🤔

Lazin replied:
removed it

nvartolomei commented on src/v/cloud_topics/core/write_pipeline.h:
```diff
@@ -0,0 +1,148 @@
+      Fn,
+      write_request<Clock>&>
+    void process_stage(Fn&& fn, pipeline_stage stage) {
+        std::set<pipeline_stage> stages;
```
nit: stages_with_new_work

Lazin replied:
what does this mean?

nvartolomei replied:
this was suggestion for a better name

nvartolomei commented on src/v/cloud_topics/core/write_pipeline.h:
```diff
@@ -0,0 +1,148 @@
+struct write_pipeline_accessor;
+
+template<class Clock = ss::lowres_clock>
+class write_pipeline {
```
missing stop() method and a vassert(_gate.is_closed(), ...) in destructor?

Lazin replied:
it's in the next PR

nvartolomei replied:
what is "next PR"?

nvartolomei commented on src/v/cloud_topics/core/write_pipeline.cc:
```diff
@@ -0,0 +1,246 @@
+  , _stages(max_pipeline_stages) {}
+
+template<class Clock>
+write_pipeline<Clock>::~write_pipeline() = default;
```
shall this (or stop()) cancel all _pending requests?

nvartolomei replied:
Or that's request responsibility?

Lazin replied:
it will not be added in this PR

nvartolomei commented on src/v/cloud_topics/throttler/throttler.cc:
```diff
@@ -0,0 +1,170 @@
+    auto list = _pipeline.get_write_requests(overshoot, _my_stage);
+    chunked_vector<write_req_ptr> tmp;
+    for (auto& wr : list.ready) {
+        _throttled.push_back(wr.weak_from_this());
```
i can't find where you are cleaning up this chunked_vector

Lazin replied:
forgot to remove `_throttled`
fixed

nvartolomei commented on src/v/cloud_topics/core/write_pipeline.h:
```diff
@@ -0,0 +1,148 @@
+    /// before back into the pipeline.
+    /// The method allows to reenqueue requests returned by get_write_requests
+    /// method.
+    void reenqueue(write_request<Clock>&);
```
The name of this method is confusing. I expected it to put the write request back at the beginning of the pipeline but after reading it and looking how throttler uses it I now understand that you have a different idea for it.

It is also a bit confusing that the last stage of the pipeline can not use it as it will catch an assert trying to get the next pipeline stage (doesn't exist).

Don't have a better idea just sharing thoughts I had as a reader.

Lazin replied:
I didn't want to change the name because the name is used in the doc that I wrote and shared with ppl. Let's not dwell on this.
